lib/abstractivator/proc_ext.rb

```diff
@@ -68,29 +68,56 @@ def proxy_call(*args, **kws, &block)
     end
   end
 
-  # tries to coerce x into a procedure, then calls it with
-  # the given argument list.
+  LooseCallInfo = Struct.new(:params, :accepts_arg_splat, :total_arity, :req_arity,
+                             :requires_kw_customization, :all_key_names, :kw_padding)
+
+  # Tries to coerce x into a procedure, then calls it with the given argument list.
   # If x cannot be coerced into a procedure, returns x.
+  # This method is optimized for use cases typically found in tight loops,
+  # namely where x is either a symbol or a keyword-less fixed-arity proc.
+  # It attempts to minimize the number of intermediate arrays created for these cases
+  # (as would be produced by calls to #map, #select, #take, #pad_right, etc.)
+  # CPU overhead created by loose_call is bad, but unexpected memory consumption would
+  # be worse, considering Proc#call has zero memory footprint.
+  # These optimizations produce a ~5x speedup, which is still 2-4x slower than
+  # regular Proc#call.
   def self.loose_call(x, args, kws={}, &block)
+    return x.to_proc.call(*args) if x.is_a?(Symbol) # optimization for a typical use case
     x = x.to_proc if x.respond_to?(:to_proc)
-    x.callable? or return x
-    arg_types = x.parameters.map(&:first)
+    return x unless x.callable?
+
+    # cache proc info for performance
+    info = x.instance_variable_get(:@loose_call_info)
```

Comment:
Good point. Caching improved benchmark performance because the benchmark reuses the proc, but it won't benefit code that doesn't reuse the proc. I'll check how much overhead caching adds for code that doesn't reuse the proc...